### PR TITLE
Reproject DataTiles

### LINKED ISSUE #13197: Reprojection for WebGlTile layer
*I'm not sure I'm on the right track here and whether this is a bug or a feature or just my lack of knowledge:*

I'm trying to show a UTM COG on an EPSG:3857 (or 4326) map, but I can't get it working. Nothing shows up for the COG, but OSM and fitting the view to the COG extent works. I assume it doesn't support automatic reprojection or is there a way to do that? A potential use case could be showing two COGs from different UTM zones on a map.

Here's a repro based on the COG example:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>Cloud Optimized GeoTIFF (COG)</title>    
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/openlayers/openlayers.github.io@master/en/v6.10.0/css/ol.css" type="text/css">
    <script src="https://cdn.jsdelivr.net/gh/openlayers/openlayers.github.io@master/en/v6.10.0/build/ol.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/proj4js/2.7.5/proj4.js"></script>
  </head>
  <body>
    <div id="map" style="height: 400px; width: 100%; border: 1px solid black"></div>
    <script>
    // Add CRS definition for COG
    proj4.defs("EPSG:32636","+proj=utm +zone=36 +datum=WGS84 +units=m +no_defs");
    ol.proj.proj4.register(proj4);
    let cogProj = ol.proj.get("EPSG:32636");
  
    // COG source
    const source = new ol.source.GeoTIFF({
      sources: [
        {
          url: 'https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/2020/S2A_36QWD_20200701_0_L2A/TCI.tif',
        }
      ],
//    projection: cogProj
    });
    
    // Create map
    const map = new ol.Map({
      target: 'map',
      layers: [
        new ol.layer.Tile({
          source: new ol.source.OSM()
        }),
        new ol.layer.WebGLTile({
          source: source
        })
      ],
      view: new ol.View({
        projection: "EPSG:4326" // CRS for map
      })
    });
    
    // Fit to bounds of COG
    source.getView().then(view => {
      // Fit to COG extent
      let fromLonLat = ol.proj.getTransform(view.projection, map.getView().getProjection());
      let extent = ol.extent.applyTransform(view.extent, fromLonLat);
      map.getView().fit(extent);
    });
    </script>
  </body>
</html>
```

### PR DESCRIPTION
**Pull request by mike-000**
Fixes #13197

Allows reprojection of multiband and raw values GeoTIFF and DataTile sources by reprojecting 3 bytes per pixel in sequence via  raster reprojection until all bytes per pixel have been reprojected, then combining the result in a new DataTile.  A `ReprojDataTile` class avoids the need for intermediate sources.

https://deploy-preview-13654--ol-site.netlify.app/en/latest/examples/geotiff-reprojection.html

To handle non-parallel projections data for areas outside the source tile grid but inside reprojected tiles with partial coverage of the source grid is returned as zeros, so sources should be configured to include alpha.  Tiles in the reprojection tile grid which are entirely outside the source tile grid have an `EMPTY` state which is handled by the renderer.

Sources are expected to have the same band count and data type at each zoom level.

A second (and more realistic use case scenario) example is added - displaying two adjacent Sentinel COGs which are in different projections due to being in different UTM zones.

https://deploy-preview-13654--ol-site.netlify.app/en/latest/examples/multiple-cogs.html

Added a static method to determine browser accuracy:
https://deploy-preview-13654--ol-site.netlify.app/en/latest/apidoc/module-ol_source_datatile#.getReprojectCapability
Gecko browsers which might interpolate a reprojection canvas even if `imageSmoothingEnabled` is false currently account for around 3.2% of browser use https://gs.statcounter.com/browser-market-share#monthly-201708-202208

### CODE REVIEW

**Comment by mike-000** [comment]
@tschaub @ahocevar This has now been fully reworked from the original experimental implementation to be a DataTile equivalent of a ReprojTile.  Would anyone like to review it?

File: src/ol/source/DataTile.js [hunk 1]
```diff
@@ -228,6 +309,115 @@ class DataTileSource extends TileSource {
...
+  const ctx2 = createCanvasContext2D(1, 1, canvasPool);
+  ctx2.imageSmoothingEnabled = false;
+  ctx2.drawImage(ctx1.canvas, 0, 0, 2, 2, 0, 0, 1, 1);
+  const data = ctx2.getImageData(0, 0, 1, 1).data;
+  releaseCanvas(ctx1);
+  canvasPool.push(ctx1.canvas);
+  releaseCanvas(ctx2);
+  canvasPool.push(ctx2.canvas);
+  return data[0] === 0 || data[0] === 255;
+}
```
Comment (tschaub) [inline src/ol/source/DataTile.js]: I don't think this is a function that we necessarily need or want to maintain in the API (better at the application level).  So I've pushed a commit that removes it.

File: src/ol/source/DataTile.js [hunk 2]
```diff
@@ -172,6 +198,61 @@ class DataTileSource extends TileSource {
...
+      const cache = this.getTileCacheForProjection(projection);
+      let tile;
+      const tileCoordKey = getKeyZXY(z, x, y);
+      if (cache.containsKey(tileCoordKey)) {
+        tile = cache.get(tileCoordKey);
+      }
+      const key = this.getKey();
+      if (tile && tile.key == key) {
+        return tile;
+      } else {
```
Comment (tschaub) [inline src/ol/source/DataTile.js]: Minor, but we can have less indentation on the happy path if we avoid these return-else.  I've pushed a commit that reworks these and will add a linter rule separately.

File: src/ol/source/DataTile.js [hunk 3]
```diff
@@ -117,6 +126,18 @@ class DataTileSource extends TileSource {
      * @type {number}
      */
     this.bandCount = options.bandCount === undefined ? 4 : options.bandCount; // assume RGBA if undefined
+
+    /**
+     * @protected
+     * @type {!Object<string, import("../tilegrid/TileGrid.js").default>}
+     */
+    this.tileGridForProjection = {};
```
Comment (tschaub) [inline src/ol/source/DataTile.js]: These can be private.  I've pushed a commit that changes this.

**Review by tschaub — APPROVED** [review]
Thanks for your efforts on this, @mike-000.  I know this will be appreciated new functionality in the library.

File: src/ol/source/DataTile.js (hunk 1 above)
Comment (mike-000) [reply src/ol/source/DataTile.js]: No problem with that.  I now have a workaround which reduces an image (by removing rows or columns) so it does not need to be downscaled but that is perhaps best left as a follow-up PR?

**Pull request merged**